### ISSUE TITLE
Using ResNet 18 model to train new image classifier

### ISSUE THREAD

**Issue by michael-golfi** [Microsoft/CNTK#582]
Hi,

I'm trying to solve a classification problem where I have several icon classes (eg: play, pause, volume, etc...). I am trying to follow CNTK's ResNet 18 example to train a classifier on my domain (currently only 9 unique classes) and then use that model from within a C# class. 

Originally, I had run into a few issues regarding this:

1. I'm trying to train on 64x64 pixel images but I'm getting a dimensioning problem from the log when I try to do so.
2. Do I need to regenerate the mean file for training a different model?

So when I edit the .ndl and .cntk files, I have updated the image sizes to be used but it seems to be that I need to change other parameters as well?

Once I have that trained, how would I go about with the mean file?

**Comment by PatrickBue** [comment]
Hi,

I modified the CNTK Resnet-18 network to take 128x128 images as input. Below is my edited ndl file - it should be relatively easy to make the changes to 64x64 pixels. Note that I decided to remove the last layer (rn4_), so that the input to the final global average pooling is of similar size as before. Alternatively (and that’s what you probably want to do since your images are even smaller) I could have removed the downscaling in that layer by setting the convolution stride to 1.

As for the xml mean file: yes, you need to generate a different one that has 64x64=4,096 number of floats. Just a global average over all pixels in the training set should suffice.

On a different note, have you thought of upscaling your images so that you can use the ResNet models that are provided by CNTK? I know this slows things down, but you wouldn't need any retraining. Another approach could be to replace the final 1k softmax layer, replace it with a 9-way softmax, and refine the network. 

Network ndl file for 128x128 images:
```
load=ndlMacros
run=DNN

ndlMacros = [
    ImageW = 128
    ImageH = 128
    ImageC = 3
    LabelDim = 100

    features = ImageInput(ImageW, ImageH, ImageC, tag = feature, imageLayout = "cudnn")
    labels = Input(LabelDim, tag = label)
    
    # Kernels width and height.
    kW = 3
    kH = 3
    # Kernel stride.
    hs = 1
    vs = 1
    
    # Initial parameter values.
    convWScale = 7.07  
    convBValue = 0

    fcWScale = 1.13
    fcBValue = 0

    scValue = 1
    
    # Batch normalization time constant.
    bnTimeConst = 32768
]

DNN=[
	#DIMENSION: 128x128x3
    conv1WScale = 0.6
    cMap1 = 64
    conv1 = ConvBNReLULayer(features, cMap1, 147, 7, 7, 2, 2, conv1WScale, convBValue, scValue, bnTimeConst)   #147 == 7x7x3
    # Max pooling
    pool1W = 3
    pool1H = 3
    pool1hs = 2
    pool1vs = 2
    pool1 = MaxNDPooling(conv1, pool1W, pool1H, pool1hs, pool1vs)
	#DIMENSION: 32x32x64   (ORIGINAL: 56x56x64)
	
    rn1_1 = ResNetNode2A(pool1, cMap1, 576, kW, kH, convWScale, convBValue, scValue, bnTimeConst)    #576 == 3x3x64
    rn1_2 = ResNetNode2A(rn1_1, cMap1, 576, kW, kH, convWScale, convBValue, scValue, bnTimeConst)
	#DIMENSION: 32x32x64   (ORIGINAL: 56x56x64)

    cMap2 = 128
    rn2_1 = ResNetNode2BInc(rn1_2, cMap2, cMap1, 576, 1152, kW, kH, convWScale, convBValue, scValue, bnTimeConst, $stride1x1$, $stride3x3$)
    rn2_2 = ResNetNode2A(rn2_1, cMap2, 1152, kW, kH, convWScale, convBValue, scValue, bnTimeConst)  
	#DIMENSION: 16x16x128  (ORIGINAL: 28x28x128)
	
    cMap3 = 256
    rn3_1 = ResNetNode2BInc(rn2_2, cMap3, cMap2, 1152, 2304, kW, kH, convWScale, convBValue, scValue, bnTimeConst, $stride1x1$, $stride3x3$)
    rn3_2 = ResNetNode2A(rn3_1, cMap3, 2304, kW, kH, convWScale, convBValue, scValue, bnTimeConst)
	#DIMENSION: 8x8x256    (ORIGINAL: 14x14x256)

    #cMap4 = 512
    #rn4_1 = ResNetNode2BInc(rn3_2, cMap4, cMap3, 2304, 4608, kW, kH, convWScale, convBValue, scValue, bnTimeConst, $stride1x1$, $stride3x3$)
    #rn4_2 = ResNetNode2A(rn4_1, cMap4, 4608, kW, kH, convWScale, convBValue, scValue, bnTimeConst)
    #rn4_3 = ResNetNode2A(rn4_2, cMap4, 4608, kW, kH, convWScale, convBValue, scValue, bnTimeConst)
	#DIMENSION: N/A        (ORIGINAL: 7x7x512)
	
    # Global average pooling
    pool2hs = 1
    pool2vs = 1
	
    #pool2W = 7
    #pool2H = 7
	#pool5 = AveragePooling(rn4_3, pool2W, pool2H, pool2hs, pool2vs, imageLayout = "cudnn")
    #ol = DnnLayer(cMap4, labelDim, pool5, fcWScale, fcBValue)   
	
    pool2W = 8
    pool2H = 8
	pool5 = AveragePooling(rn3_2, pool2W, pool2H, pool2hs, pool2vs, imageLayout = "cudnn")
	#DIMENSION: 1x1x256    (ORIGINAL: 1x1x256)
	
    ol = DnnLayer(cMap3, labelDim, pool5, fcWScale, fcBValue)
	#DIMENSION: 1x1x100    (ORIGINAL: 1x1x1000)
    
    CE = CrossEntropyWithSoftmax(labels, ol, tag = Criteria)
    Err = ErrorPrediction(labels, ol, tag = Eval)
    OutputNodes = ol
]

```

**Comment by PatrickBue** [comment]
One more question: how many training examples do you have?

**Comment by michael-golfi** [comment]
I have about 1125 unique training images. I scaled them anyways to 64x64 from more or less random image sizes. I also greyscaled them all and then applied 16 different filters to them, so right now I have a total of 18,000 images for training.

**Comment by PatrickBue** [comment]
1125 unique training examples for 9 classes doesn't seem nearly enough to train a ResNet. You can try, but my recommendation would be to train / use a pre-trained ResNet first and either (i) train a SVM on the output of the 2nd to last layer, or (ii) replace the last layer with a 9-way softmax and then run a few training iterations.

**Comment by PatrickBue** [comment]
Here's a link that explains how to replace and retrain the last layer of a CNTK model:
https://github.com/Microsoft/CNTK/issues/430

**Comment by michael-golfi** [comment]
I retrained the ResNet 18 example using the following .cntk and .mel:
```
RootDir = "."

ConfigDir = "$RootDir$"
DataDir = "$RootDir$/Data"
OutputDir = "$RootDir$/Output"
ModelDir = "$OutputDir$/Models"

ndlMacros="$ConfigDir$/Macros.ndl"

precision="float"
deviceId="Auto"

command=Retrain:Test

parallelTrain="false"

stderr="$OutputDir$/ResNet"
traceLevel=1
numMBsToShowResult=500

# Strides for increasing layers. Defaults (paper) are 2 for 1x1 and 1 for 3x3.
stride1x1=2
stride3x3=1

Retrain=[    
    action="edit"
    CurModel="$ModelDir$/ResNet.model"
    NewModel="$ModelDir$/ResNet.Eval"
    editPath="$ConfigDir$/Retrain.mel"
]
```
```
#load pre-trained model
 model1 = LoadModel("$ModelDir$/ResNet_18.model")
 SetDefaultModel(model1)
 #DumpModel(model1, "$RootDir$/dump_originalModel.txt", includeData =true)

#add new final layer
 inDim=512
 labelDimNew=9
 fcWScale = 1.13
 fcBValue = 0
 
 newL = DnnLayer(inDim, labelDimNew, pool5, fcWScale, fcBValue)
 #newL = DNNLayer(inDim, labelDimNew, h1.y, wScale, fcBValue)
 labelsNew = Input(labelDimNew, tag="label")
 SetInput(ce, 0, labelsNew)
 SetInput(ce, 1, newL.z)
 SetInput(err, 0, labelsNew)
 SetInput(err, 1, newL.z)
 SetProperty(newL.z, "output", "true")

#remove old final layer (note:make sure these deletes happen in reverse order from the original definition)
 DeleteNode(OutputNodes.z) 
 DeleteNode(OutputNodes.t)
 DeleteNode(OutputNodes.b)
 DeleteNode(OutputNodes.W)
 DeleteNode(labels)

#rename nodes to have same name as before (so that the same .cntk training script can be used)
 Rename(labelsNew, labels)
 Rename(newL.z, OutputNodes.z)

#save and dump model
 SaveModel(model1, "$ModelDir$/ApeNet_18.model", format="cntk")
 #DumpModel(model1, "$RootDir$/dump_editedModel.txt", includeData =true)
```

Now, I am trying to evaluate on some of the training images to see if its working but I'm getting the following values (the test image is part of class 4):

```
		[0]	-0.201454237	float
		[1]	-0.10742113	float
		[2]	-0.183760867	float
		[3]	0.141525343	float
		[4]	-0.0753818154	float
		[5]	-0.0116478	float
		[6]	-0.0975382	float
```

**Comment by PatrickBue** [comment]
The negative values are to be expected, since the softmax is done in the loss function. See here regarding training the last layer you added:
https://github.com/Microsoft/CNTK/issues/430

**Comment by michael-golfi** [comment]
When I uncomment DumpModel, the training fails to complete with the following error:

`EXCEPTION occurred: Node 'newL.W' (LearnableParameter operation): Sample [7 x 1 x 1 x 512] is not a column vector or matrix (1D or 2D tensor).

[CALL STACK]
    > Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - 00007FF6505F3793 (SymFromAddr() error: Attempt to access invalid address.)
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - Microsoft::MSR::CNTK::MatrixBase::  operator=
    - BaseThreadInitThunk
    - RtlUserThreadStart`

When I run it without DumpModel it completes.

**Comment by PatrickBue** [comment]
The DumpModel function might be broken, I also got that error. See: https://github.com/Microsoft/CNTK/issues/438

**Comment by michael-golfi** [comment]
So I've managed to start training the model, seemingly successfully. But it crashes at some point after starting the training. I get no error message though... I'm a bit stumped at this and I've been trying to work through this for a couple of days. I set the verbosity on the reader to 2.

`Allocating matrices for forward and/or backward propagation.
No PreCompute nodes found, skipping PreCompute step.

Starting Epoch 1: learning rate per sample = 0.004688  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 1 of 125]: [Training Set] TrainLossPerSample = 0.026576127; TotalSamplesSeen = 5056; EvalErrPerSample = 0.012658228; AvgLearningRatePerSample = 0.0046875002; EpochTime=63.8566
SGD: Saving checkpoint model './Output/Models/ApeNet.model.1'

Starting Epoch 2: learning rate per sample = 0.000469  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 2 of 125]: [Training Set] TrainLossPerSample = 10.094723; TotalSamplesSeen = 10112; EvalErrPerSample = 0.16178797; AvgLearningRatePerSample = 0.00046874999; EpochTime=57.6492
SGD: Saving checkpoint model './Output/Models/ApeNet.model.2'

Starting Epoch 3: learning rate per sample = 0.000047  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 3 of 125]: [Training Set] TrainLossPerSample = 0.00015859134; TotalSamplesSeen = 15168; EvalErrPerSample = 0; AvgLearningRatePerSample = 4.6875e-005; EpochTime=57.7199
SGD: Saving checkpoint model './Output/Models/ApeNet.model.3'

Starting Epoch 4: learning rate per sample = 0.000047  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 4 of 125]: [Training Set] TrainLossPerSample = 8.8424929e-005; TotalSamplesSeen = 20224; EvalErrPerSample = 0; AvgLearningRatePerSample = 4.6875e-005; EpochTime=57.7041
SGD: Saving checkpoint model './Output/Models/ApeNet.model.4'

Starting Epoch 5: learning rate per sample = 0.000047  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 5 of 125]: [Training Set] TrainLossPerSample = 0.0001424779; TotalSamplesSeen = 25280; EvalErrPerSample = 0; AvgLearningRatePerSample = 4.6875e-005; EpochTime=57.677
SGD: Saving checkpoint model './Output/Models/ApeNet.model.5'

Starting Epoch 6: learning rate per sample = 0.000047  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
Finished Epoch[ 6 of 125]: [Training Set] TrainLossPerSample = 0.00013595171; TotalSamplesSeen = 30336; EvalErrPerSample = 0; AvgLearningRatePerSample = 4.6875e-005; EpochTime=57.6739
SGD: Saving checkpoint model './Output/Models/ApeNet.model.6'

Starting Epoch 7: learning rate per sample = 0.000047  effective momentum = 0.900000  momentum as time constant = 607.4 samples

Starting minibatch loop.
`

**Comment by eldakms** [comment]
Hi Michael,

Could you please provide more information about your environment:
 - which operating system?
 - are you on the latest master (1.5 binary drop)? Thanks!
Thank you!

**Comment by michael-golfi** [comment]
Hi @eldakms ,

Currently I'm running on Windows 10
I'm using the latest binary. 

I was first training on a NVS 310 but @frankseide mentioned I should be using a GPU with at least compute capability > 3.0 due to an Nvidia bug. I now train on a Tesla K40. I still see the same issue.

I have access to a Quadro M6000, I can try to train on that soon.

**Comment by michael-golfi** [comment]
I ran it through debug after CNTK.exe crashes and ended up with this message:

Unhandled exception at 0x00007FFB8E6C4A30 (msvcr120.dll) in CNTK.exe: Fatal program exit requested.

**Comment by eldakms** [comment]
Hi Michael,

I suppose you use the correct version of OpenCV, right?
Could you please use procdump, collect full dump and share it with me?
If it is not easily reproducible you can enable procdump to collect the traces in the passive mode (https://technet.microsoft.com/en-gb/sysinternals/dd996900.aspx, i.e. 
    Register as the Just-in-Time (AeDebug) debugger. Makes full dumps in c:\dumps.
    C:\>procdump -ma -i c:\dumps
) 

Thank you very much!

**Comment by michael-golfi** [comment]
@eldakms I think I figured it out earlier this morning. I believe CNTK crashes when the image file names are malformed, but it doesn't ouput any indication of the error cause.

I'm going to try and repro it.

**Comment by eldakms** [comment]
If this is the case, could you please describe how I can reproduce this and I will fix this asap. Thank you very much for digging into this!

**Comment by eldakms** [comment]
Image reader is currently implemented using OpenCV. Unfortunately OpenCV does not support Unicode file names on Windows (see here https://github.com/Itseez/opencv/issues/4292 )
I have added more diagnostics to the output and fixed the AV. Thank you very much for reporting this! I will update the issue as soon as the fix is in master

**Comment by michael-golfi** [comment]
Hi @eldakms,

I'm going to move to close since my stuff is now working based on correcting the file paths. So in conclusion, the file names should not be Unicode. The following code should help as a temporary solution. In order to validate the string for Unicode characters:

`public static bool IsUnicode(string input)
{
      const int MaxAnsiCode = 255;
      return input.ToCharArray().Any(c =&gt; c &gt; MaxAnsiCode);
}
`

To remove Unicode characters:

`
public static String StripUnicodeCharactersFromString(string inputValue)
{
	return new string(inputValue.Where(c => c <= sbyte.MaxValue).ToArray());
}`

Or if you don't like LinQ:

`
public static String StripUnicodeCharactersFromString(string inputValue)
{
	return Encoding.ASCII.GetString(Encoding.Convert(Encoding.UTF8, Encoding.GetEncoding(Encoding.ASCII.EncodingName, new EncoderReplacementFallback(String.Empty), new DecoderExceptionFallback()), Encoding.UTF8.GetBytes(inputValue)));
}`

@eldakms can we close?
